src/Npgsql.Json.NET/Internal/JsonHandler.cs

```diff
@@ -27,60 +28,61 @@ class JsonHandler : Npgsql.Internal.TypeHandlers.JsonHandler
         public JsonHandler(PostgresType postgresType, NpgsqlConnector connector, JsonSerializerSettings settings)
             : base(postgresType, connector, isJsonb: false) => _settings = settings;
 
-        protected override async ValueTask<T> Read<T>(NpgsqlReadBuffer buf, int len, bool async, FieldDescription? fieldDescription = null)
+        protected override async ValueTask<T> ReadCustom<T>(NpgsqlReadBuffer buf, int len, bool async, FieldDescription? fieldDescription = null)
         {
             if (typeof(T) == typeof(string) ||
                 typeof(T) == typeof(char[]) ||
                 typeof(T) == typeof(ArraySegment<char>) ||
                 typeof(T) == typeof(char) ||
                 typeof(T) == typeof(byte[]))
             {
-                return await base.Read<T>(buf, len, async, fieldDescription);
+                return await base.ReadCustom<T>(buf, len, async, fieldDescription);
             }
 
             // JSON.NET returns null if no JSON content was found. This means null may get returned even if T is a non-nullable reference
             // type (for value types, an exception will be thrown).
             return JsonConvert.DeserializeObject<T>(await base.Read<string>(buf, len, async, fieldDescription), _settings)!;
         }
 
-        protected override int ValidateAndGetLength<T2>(T2 value, ref NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter)
+        protected override int ValidateAndGetLengthCustom<T2>([DisallowNull] T2 value, ref NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter)
```

Comment:
Instead of changing the custom methods to have "Custom" postfixed we could inverse it. Even though the api is 'internal' there will be users that will be broken by this.

Reply:
Good point, though we've already broken users by moving all the types into the Internal namespace... So at this point I'd rather just have nice naming...